packages/promise-cache/src/promiseCache.ts

```diff
@@ -7,11 +7,19 @@ export class PromiseCache<T, U> {
     this.ttl = ttlInSeconds * 1000 // Convert seconds to milliseconds
   }
 
-  async wrap(key: T, delegate: () => Promise<U>): Promise<U> {
+  async wrap(
+    key: T,
+    delegate: () => Promise<U>,
+    ttlInSeconds?: number
+  ): Promise<U> {
     const now = Date.now()
     const cached = this.cache.get(key)
 
-    if (cached && now - cached.timestamp < this.ttl) {
+    // Determine the TTL for this specific call
+    const effectiveTTL =
+      ttlInSeconds !== undefined ? ttlInSeconds * 1000 : this.ttl
+
+    if (cached && now - cached.timestamp < effectiveTTL) {
```

Comment:
Wouldn't it be better to use setTimeout with ttl and remove the key/val from the map? Or at least remove the value when it is detected that it is expired?

If using the timeout route, you would of course need to save the timeout identifier so that you can clear the timeout if a new value is cached with the same key while the timeout is running

Reply:
Yes good point!

I thought about it for a while and mixing cache calls with the same key but different TTL's is weird behavior on the consumer side.

We would need to calculate the difference between the TTL's and update the cache and/or recreate the setTimeout with a new value. I also thought about including the TTL in the cache key itself... but that can confuse consumers even more. It gets pretty messy...

So I ended up adding ttl to the cached object and a console.error call that warns if the same key has been used with different TTL's. What do you think?